### ISSUE TITLE
Is there a way to catch the result of a query with relations after they're resolved?

### ISSUE THREAD

**Issue by tonyketcham** [graphql-compose/graphql-compose#368]
I would like to wait for relations and child resolvers to completely resolve first, then run some code based on query args to modify the result before its sent to the user. 

Reading through all the documentation, `withMiddlewares()` seemed to be what I was looking for. However, using it to wrap my `all<type>` query resolvers like so:
```js
schemaComposer.Query.addFields({
  /**
   * Add find by ID to each content type
   */
  [type]: schema.getResolver('findById'),
  /**
   * Add find 'many' to each content type
   */
  [pluralTypeQueryName]: schema
    .getResolver('all')
    .withMiddlewares([logResultOfResolver, cascadingFilterMiddleware]),
});

...

const logResultOfResolver = async (
  resolve: (arg0: any, arg1: any, arg2: any, arg3: any) => any,
  source: any,
  args: any,
  context: any,
  info: any
) => {
  const result = await resolve(source, args, context, info);
  console.log('result', result);
  return result;
};
```

`result` in `logResultOfResolver` contains the initial unresolved data instead of the resolved relations:
```js
result [
  {
    filename: 'anotha-one.md',
    path: 'content\\markdown\\posts\\anotha-one.md',
    slug: 'anotha-one',
    id: '92348fds-453fdh-59ddsd-3332-09876',
    title: 'Senior cat list',
    authors: [ '40s3', '2a3e' ],
    rating: 100,
    content: { raw: '\n**all of them are good**\n' }
  },
  {
    filename: 'example-post.md',
    path: 'content\\markdown\\posts\\example-post.md',
    slug: 'example-post',
    id: 'sdfsdf-23423-sdfsd-23444-dfghf',
    title: 'Example post of things',
    authors: [ '2a3e', '40s3' ],
    rating: 74,
    content: {
      raw: '\n' +
        '# My cat is so little\n' +
        '\n' +
        '## she is so small\n' +
        '\n' +
        '### she is in fact\n' +
        '\n' +
        '2 apples tall\n'
    }
  }
]
```

However the actual query returns the correctly resolved response to the client:
```js
{
  "data": {
    "allPosts": [
      {
        "title": "Senior cat list",
        "authors": [
          {
            "name": "Eva",
            "id": "40s3"
          },
          {
            "name": "Tony",
            "id": "2a3e"
          }
        ]
      },
      {
        "title": "Example post of things",
        "authors": [
          {
            "name": "Eva",
            "id": "40s3"
          },
          {
            "name": "Tony",
            "id": "2a3e"
          }
        ]
      }
    ]
  }
}
```

## Some context
In case I'm just using the wrong approach entirely, what I'm trying to accomplish is allowing users of my library to pass in filter queries of the following syntax:

```graphql
query FilteredPosts {
  allPosts(
    filter: { title: { wildcard: "*tion" }, postMeta: { rating: { gt: 80 } } }
  ) {
    title
  }
}
```
```js
entries = [
  { id: 1, title: 'My pretzel collection', postMeta: { rating: 97 } },
  { id: 2, title: 'Debugging the simulation', postMeta: { rating: 20 } },
  {
    id: 3,
    title: 'Liquid Proust is a great tea vendor btw',
    postMeta: { rating: 99 },
  },
  { id: 4, title: 'Sitting in a chair', postMeta: { rating: 74 } },
];
```

Resulting in:

```js
result = [{ id: 1, title: 'My pretzel collection', postMeta: { rating: 97 } }];
```

I apologize for the long issue and would greatly appreciate any guidance on how to solve this problem -- it's been eating me alive for like 3 days hahaha

Fantastic work on this library, by the way! 😊

**Comment by tonyketcham** [comment]
After spelunking, I found that [this thread](https://github.com/graphql-compose/graphql-compose-mongoose/issues/96) goes into quite a bit of detail on the subject. It seems I would need to create custom filter resolvers that may operate outside of GraphQL and GraphQL-compose. 

Though for filtering on nested fields which must first be resolved based on underlying node data (ex. content exists but timeToRead must be resolved based on the content) I'm wondering if there's a way to make use of existing resolvers or if this would just be entirely custom outside of the flow of GraphQL-compose.

I could just use the underlying functions of the resolvers and bake everything that the filter depends on, then generate a list of posts, for example, that have the IDs that the filter generates. From there, the GraphQL flow could take over and proceed as normal. 

It seems to me like this approach would resolve many fields twice, which may necessitate some memoization to cut down on overhead.

**Comment by tonyketcham** [comment]
To keep that custom filter resolver DRY, I may be able to utilize the existing field resolvers from the corresponding ObjectTCs by traversing through the collection types and checking if a field is a relation to another collection or not.

**Comment by nodkz** [comment]
Yep, you need to write your custom resolver. It's almost impossible to generate well-adopted filters for custom cases. And if you have specific requirements - it's better to write your resolver manually.

Moreover, if you want to create a complex filter that works with nested relations - it definitely requires writing a custom resolver that fetches all required data from database in a top-level resolver using provided filter arguments.

Under the hood, GraphQL starts requesting nested data only after the execution of the top-level resolver. It resolves 1st level fields after that starts resolving 2nd level fields, after that 3rd and so on. So if you need "LEFT JOIN" filtering the only place where it can be done is the top-resolver. You need to prepare all nested data set in your resolve function which match GraphQL query shape and just return it.
